tools/diff-kernel-config

```diff
@@ -0,0 +1,241 @@
+                        build-package \
+                    || bail "Build failed for ${debug_id}"
+
+                #
+                # Find kernel RPM
+                #
+
+                shopt -s nullglob
+                kernel_rpms=(
+                    ./build/rpms/bottlerocket-"${arch}"-*kernel-"${kver}"-"${kver}".*.rpm
+                )
+                shopt -u nullglob
+
+                case ${#kernel_rpms[@]} in
+                    0) bail "No kernel RPM found for ${debug_id}" ;;
+                    1) kernel_rpm=${kernel_rpms[0]} ;;
+                    *) bail "More than one kernel RPM found for ${debug_id}" ;;
+                esac
+
+
+                #
+                # Extract kernel config
+                #
+
+                config_path=./${output_dir}/config-${arch}-${kver}-${variant}-${state}
+                rpm2cpio "${kernel_rpm}" \
+                    | cpio --quiet --extract --to-stdout ./boot/config >"${config_path}"
+                [[ -s "${config_path}" ]] || bail "Failed to extract config for ${debug_id}"
+
+            done  # variant
+
+        done  # kver
+
+    done  # arch
+
+done  # state
+
+
+#
+# Post-process the collected pairs of "before" and "after" configs (generate diffs, a report, a summary)
+#
+
+# Get the helpful diffconfig script from the kernel source tree. We package it
+# in the kernel-archive RPM from where it can be extracted. Here we extract the
+# latest version of the script, but any kernel version and arch will do.
+latest_kver=$(printf '%s\n' "${kernel_versions[@]}" | sort -V | tail -n1)
+latest_archive_rpms=( ./build/rpms/bottlerocket-aarch64-kernel-"${latest_kver}"-archive-*.rpm )
+diffconfig=$(mktemp --suffix -bottlerocket-diffconfig)
+rpm2cpio "${latest_archive_rpms[0]}" \
+    | cpio --quiet --extract --to-stdout \
+    | tar --xz --extract --to-stdout kernel-devel/scripts/diffconfig >"${diffconfig}"
+[[ -s ${diffconfig} ]] || bail "Failed to extract diffconfig tool from '${latest_archive_rpms[0]}'."
+chmod +x "${diffconfig}"
+
+# Diff the before and after states for each collected pair
+for config_before in "${output_dir}"/config-*-before; do
+    config_after=${config_before/before/after}
+    config_diff=${config_before/before/diff}
+    "${diffconfig}" "${config_before}" "${config_after}" >"${config_diff}" \
+        || bail "Failed to diff '${config_before}' and '${config_after}'"
+done
+
+# Generate diff summary
+echo
+for config_diff in "${output_dir}"/config-*-diff; do
+    config_base=${config_diff##*/}
+    awk "
+        /^-/   { removed += 1 }
+        /^+/   { added   += 1 }
+        / -> / { changed += 1 }
+        END    { printf \"${config_base}:\t%3d removed, %3d added, %3d changed\n\", removed, added, changed }
+    " "${config_diff}"
+done | sort -V
+echo
+
+# Generate combined report of changes
+head -v -n 999999 "${output_dir}"/*-diff >"${output_dir}"/diff-report
+echo "A full report has been placed in '${output_dir}/diff-report'"
```

Comment:
If the report itself is the only reason to keep the directory around, it might be a nicer interface to use a temporary directory for the output, and just copy the report out as a permanent, gitignore'd file at the end.

That would trade off the need to keep track of output directories that don't yet exist with the need to occasionally prune older report files.

Reply:
Personally, I haven't really made up my mind yet on this matter. I've found use for both the combined report and the individual files. Let's see how folks will use it in the next PRs and decide one way or the other? I agree on simplifying the interface should it turn out the combined report is all that's getting used.